[카프카] 로또 1단계 - 자동 미션 제출합니다

include42 opened the pull request:
안녕하세요. 카프카입니다.

지적해주신 내용을 정리하면서 다시금 반성하는 부분도 있고, 고민한 부분도 있었습니다.
특히 LottoRank 내의 calculateRank 메서드는 설계하면서도 많이 고민했던 부분인데,
지적해주신 내용을 바탕으로 수정하면서 나름의 해답을 찾아 기쁘게 생각하고 있습니다.

지적해주신 내용 위주로 고치고 몇 가지 자잘한 수정을 했습니다만,
아직 많이 부족한 만큼 혹여나 잘못된 부분을 보신다면 지적 부탁드리겠습니다.
리뷰 보내주시는대로 확인하겠습니다.
감사합니다.

include42 committed:
[docs] README에 요구사항 및 구현단위 작성

README 요구사항을 간략히 정리하고 테스트를 위한 구현 단위 일부 작성

include42 committed:
[feat] Money 클래스 구현 및 MoneyTest 구현

Money 로또 구입 로직에서 금액을 담당
MoneyTest Money에 대한 테스트(생성, 입력, 예외처리) 구현

include42 committed:
[refactor] Money 클래스 refactor

Money 예외처리 메소드 분리 및 상수 처리

include42 committed:
[docs] README 구현단위 작성

README 로또번호 클래스 관련 테스트단위 작성

include42 committed:
[feat] LottoNumber 클래스 구현

LottoNumber 생성자 및 equals 로직 구현
LottoNumberTest 생성, 예외처리, 비교연산 테스트 수행

include42 committed:
[feat] LottoNumbers 클래스 구현

LottoNumbers 생성자 및 공통 요소 검사 기능
LottoNumbersTest 생성, 예외처리, 공통요소 count 테스트 수행

include42 committed:
[refactor] LottoNumbers 상수 처리

include42 committed:
[feat] NumberGenerator 인터페이스 구현 및 RandomNumberGenerator 구현

NumberGenerator LottoNumbers 생성을 위한 인터페이스
RandomNumberGenerator LottoNumbers를 랜덤으로 생성하기 위한 클래스

include42 committed:
[feat] LottoNumbersFactory 클래스 구현

LottoNumbersFactory create 메소드를 NumberGenerator를 받아 LottoNumbers 생성
LottoNumbersFactoryTest 생성테스트 구현

include42 committed:
[feat] LottoRank enum 구현

LottoRank 게임 결과값을 이용해 Rank 생성
LottoRankTest Rank 생성 테스트

include42 committed:
[feat] LottoRank enum 구현

LottoRank 게임 결과값을 이용해 Rank 생성

include42 committed:
[feat] LottoWinner 클래스 구현 및 코드 리팩토링

LottoWinner 당첨번호와 다른 Numbers를 비교하는 클래스 구현
LottoWinnerTest 생성 및 랭크 반환 테스트 수행
ALL 컨벤션 준수를 위한 코드 리팩토링

include42 committed:
[docs] README 구현단위 작성

README 열거형 및 당첨번호 관련하여 구현내용 작성

include42 committed:
[feat] LottoResult 클래스 구현 및 코드 리팩토링

LottoResult 당첨 결과를 저장하는 클래스 구현
LottoResultTest LottoResult의 생성, 입출력, 예외 테스트 검증
ResultCount LottoResult 내의 당첨 횟수 기록을 담당하는 클래스
ResultCountTest ResultCount의 생성, 증가 메서드 검증

include42 committed:
[feat] 뷰 클래스 작성

InputView 입력을 담당하는 뷰 클래스 구현
OutputView 출력을 담당하는 뷰 클래스 구현

include42 committed:
[feat] 컨트롤러 클래스 작성 및 작동 확인

LottoController 컨트롤러 역할을 하는 클래스 구현, 작동 확인
Application 메인 로직을 수행하는 클래스
ALL MVC 구조에 맞춰 코드 일부 리팩토링

include42 committed:
[refactor] LottoNumber 클래스 리팩토링

LottoNumber 캐시값을 통해 객체 재활용
ALL LottoNumber 구조변경에 따른 리팩토링

include42 committed:
[refactor] 패키지 리팩토링

ALL 패키지 리팩토링

include42 committed:
[refactor] 코드 리팩토링

ALL 컨벤션 준수를 위한 코드 리팩토링

include42 committed:
[refactor] 정적 클래스의 생성자 호출 방지 구현

LottoNumbersFactory,Input,Output 생성자 호출 시 예외 발생

include42 committed:
[refactor] LottoController 내 메서드 분리

LottoController 기능에 따라 메서드 분리

include42 committed:
[refactor] 테스트코드 내 코드 리팩토링

TEST 테스트코드 내의 컨벤션 미준수에 따른 코드 리팩토링

include42 committed:
[refactor] UserNumberGenerator 리팩토링

UserNumberGenerator init 메서드 재설계(메모리 절약을 위해 clear 후 재활용)

include42 committed:
[refactor] 코드 컨벤션 맞춤

include42 committed:
[refactor] 요구사항에 맞춰 OutputView 기능 추가

LottoRank bonus getter 추가
LottoController OutputView에 로또게임 횟수 및 보너스 여부 전달
OutputView LottoController의 수정사항에 맞춰 로직 변경

include42 committed:
[refactor] LottoNumbers 내의 List SortedSet으로 변경

LottoNumbers 내부 인자 중복을 막고 정렬을 생략하기 위해 SortedSet 사용
ALL LottoNumbers 변경에 따른 코드 리팩토링

include42 committed:
[refactor] LottoResult 내의 HashMap을 EnumMap으로 변경

include42 committed:
[feat] LottoGame 클래스 구현

LottoGame List<LottoNumbers>를 감싸는 1급콜렉션 역할을 수행하는 클래스 구현
LottoResult 수익 계산 메서드 추가
ResultCount 문서화 주석 추가
OutputView 출력 메서드 재설계. LottoResult를 받아 printResult 메서드 수행

include42 committed:
[refactor] NumberGenerator 이름을 LottoGenerator로 변경

NumberGenerator 삭제
LottoGenerator 이름 변경에 따른 생성, 상속받은 클래스 모두 이름 변경.
ALL 컨벤션 및 이름변경에 따른 리팩토링

hongsii commented on src/main/java/domain/lottonumber/LottoNumbers.java:
```diff
@@ -0,0 +1,43 @@
+    private static final String ERROR_INVALID_SIZE_MESSAGE = "인자의 갯수가 올바르지 않습니다.";
+    private static final int LOTTO_SIZE = 6;
+
+    private SortedSet<LottoNumber> lottoNumbers;
```
`SortedSet` 활용 👍

hongsii commented on src/main/java/domain/lottonumber/generator/UserLottoGenerator.java:
```diff
@@ -0,0 +1,21 @@
+import java.util.*;
+
+public class UserLottoGenerator implements LottoGenerator {
+    SortedSet<LottoNumber> numbers = new TreeSet<>();
```
인스턴스 변수는 정말정말 특별한 이유가 없다면 `private` 으로 선언해 외부로 노출되는 것을 막아주세요.
외부로 노출된 내용이 이곳저곳에서 변경할 수 있기 때문에 최대한 내부에서만 상태를 변경할 수 있게 제한하면 외부의 사용에 상관없이 클래스 수정이 쉽습니다.

include42 replied:
급하게 짜면서 빠뜨렸네요!! 이런 실수 없도록 신경써야겠습니다ㅎㅎ

hongsii commented on src/main/java/domain/lottonumber/generator/UserLottoGenerator.java:
```diff
@@ -0,0 +1,21 @@
+    SortedSet<LottoNumber> numbers = new TreeSet<>();
+
+    public void init(List<Integer> inputNumbers) {
+        numbers.clear();
```
멀티 쓰레드 환경에서도 동작하기 위해선 최대한 불변 객체로 코딩하도록 연습하시면 좋을 것 같습니다. 인스턴스를 재사용하지 않고 유저 입력마다 매번 생성하도록 바꿔보면 어떨까요?

include42 replied:
필드 변수를 최대한 활용하는게 좋다고 생각했는데, 말씀해주신 내용을 보니 불변성에 대한 부분을 놓쳤네요!!
이 부분은 말씀주신 내용을 적용하여 개선했습니다~

hongsii commented on src/main/java/domain/lottoresult/LottoRank.java:
```diff
@@ -0,0 +1,48 @@
+        return Arrays.stream(LottoRank.values())
+                .filter(rank -> rank.bonus == bonus && rank.hitCount == hitCount)
+                .findAny()
+                .orElse(calculateRankWithoutBonus(hitCount));
```
`orElse()` 는 값이 없더라도 이 메서드가 실행은 됩니다. 상수가 아닌 메서드를 매번 호출한다면 `orElseGet()` 을 활용하면 좋을 것 같습니다. 

- https://stackoverflow.com/questions/33170109/difference-between-optional-orelse-and-optional-orelseget

include42 replied:
이렇게 작동하는 줄은 몰랐네요!! 다시 공부해볼 욕심이 납니다ㅎㅎ

hongsii commented on src/test/java/domain/lottonumber/LottoNumbersTest.java:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    @Test
+    void 인자가_6개가_아니면_예외처리() {
```
테스트 케이스만 다른 경우면 `@ParameterizedTest` 를 활용하면 테스트 코드의 가독성을 높일 수 있습니다.

include42 replied:
테스트 코드들이 리팩토링이 잘 안되어서, 말씀주신 내용 보고 많이 수정했습니다.
좀 더 효율적으로 짜봐야겠어요!!

hongsii commented on src/main/java/view/InputView.java:
```diff
@@ -0,0 +1,48 @@
+import java.util.Scanner;
+import java.util.stream.Collectors;
+
+public class InputView {
```
> 뷰의 역할이 단순히 입력/출력에만 머무르는게 옳은 것인가요?
> 아니면 입출력 데이터를 수정하거나 예외처리를 담당하거나 해도 괜찮은가요?

이건 상황과 요구사항, 설계에 따라 다르지만, 도메인 로직을 실행하기 위한 데이터 조작은 뷰에서 처리해도 무방하다고 생각합니다.

include42 replied:
뷰의 역할에 대해 너무 강박적으로 접근했던 것 같습니다. 조언 주셔서 감사합니다!!

hongsii reviewed (changes requested):
안녕하세요, 리뷰를 맡게 된 홍시입니다.
요구사항 맞게 전체적으로 잘 구현해주셨네요!
피드백 확인하셔서, 반영해주세요~

include42 committed:
[refactor] 수정사항 반영에 따른 코드 리팩토링

LottoNumbersTest ParameterizedTest로 개수 테스트 메서드 변경
LottoRank 순위 계산 메서드의 stream 오류 개선(Else -> ElseGet)
UserLottoGenerator 객체 재활용 로직을 새로 생성하는 로직으로 수정

include42 committed:
[refactor] OutputView 메서드 병합 및 LottoController 리팩토링

OutputView printResultTitle과 printEarning을 printResult에 통합
LottoController LottoResult를 필드변수에서 run의 지역변수로 변경

include42 committed:
[refactor] LottoNumbers 클래스 LottoTicket으로 이름 변경

LottoNumbers 삭제
LottoTicket 기존의 LottoNumbers 역할을 수행
LottoNumbersFactory 삭제
LottoTicketFactory 기존의 LottoNumbersFactory 역할 수행
ALL 이름 변경에 맞춘 코드 리팩토링

include42 committed:
[refactor] LottoNumberFactory 구현

LottoNumberFactory LottoNumber를 캐싱해두고 필요할 때 반환
LottoNumber 캐싱 및 of 메소드 기능 LottoNumberFactory로 이동
LottoNumberFactoryTest Factory를 통한 번호생성, 값비교, 예외 검증

include42 committed:
[refactor] 컨벤션 준수를 위한 리팩토링

LottoNumberFactory 코드 리팩토링

include42 committed:
[refactor] LottoTicketTest 중복코드 메서드 분리

LottoTicketTest 중복되는 생성 코드를 메서드로 분리

include42 committed:
[refactor] LottoNumberFactory 내 메소드 이름변경 및 테스트 리팩토링

LottoRankTest 반복문을 통해 중복 코드 제거
LottoNumberFactory 메소드 getInstance를 getLottoNumber로 변경
ALL 메소드명 변경에 따른 코드 리팩토링

include42 committed:
[refactor] InputView 내 에러메세지 상수화

InputView NumberFormat 관련 예외처리 메세지 상수화

include42 committed:
[refactor] LottoResult 메소드명 변경 및 LottoNumberFactory 생성자 구현

LottoNumberFactory private 생성자 추가하여 생성 방지
LottoResult calculateEarning 이름 calculateGamesEarning으로 변경

hongsii commented on src/main/java/domain/lottoresult/LottoRank.java:
```diff
@@ -0,0 +1,48 @@
+        return Arrays.stream(LottoRank.values())
+                .filter(rank -> rank.bonus == bonus && rank.hitCount == hitCount)
+                .findAny()
+                .orElseGet(() -> calculateRankWithoutBonus(hitCount));
```
개수와 보너스가 일치하지 않으면 바로 `NOTHING` 을 반환하는 게 더 적절하지 않을까요?
`bonus` 를 구하는 로직에 버그가 생긴다면 문제가 될 수 있는 로직으로 보입니다.

include42 replied:
이 부분에 대해서도 고민이 많았습니다..

로또 2등을 제외하면 보너스 여부와 상관 없이 당첨 개수(hitCount)가 맞는 값이 반환되어야 하는데,
그렇다고 2등만 명시해서 예외적으로 검사하면 이상할 듯 해서 페어와 상의를 했습니다.

결과적으로는 보너스+당첨번호 모두 일치(로또 2등(5개일치 및 보너스 일치)처럼 
보너스가 값에 영향을 주는 경우)를 먼저 검사하고, 찾지 못할 경우 
보너스와 상관없는 결과(당첨번호 개수만 같은, 
로또1등(6개일치)나 로또3등(5개일치) 같은)를 찾는 메서드를 다시 호출하여 값을 반환하게 하였습니다.

이 부분은 아예 새로 설계해서 다시 리뷰를 보내드리겠습니다.

include42 replied:
메서드를 호출하지 않아도 작동하도록 수정하였습니다!!
내용의 이해를 위해 메서드 내에 문서화주석 첨부하였습니다.

hongsii commented on src/main/java/domain/lottonumber/generator/UserLottoGenerator.java:
```diff
@@ -0,0 +1,23 @@
+public class UserLottoGenerator implements LottoGenerator {
+    private SortedSet<LottoNumber> numbers;
+
+    public void init(List<Integer> inputNumbers) {
```
해당 클래스를 재사용하는 경우에 동시성 문제가 발생할 수 있습니다. 해당 로직을 생성자로 변경하면 불변 객체로 만들 수 있습니다.

include42 replied:
안그래도 init라는 메서드가 선행되어야만 create가 정상적으로 작동한다는 구조적 문제가 있어 고민이었습니다.
생성을 최대한 줄여서 메모리를 아끼는게 좋지 않을까? 라고 생각했었는데 문제점이 더 많았네요. 좋은 지적 감사드립니다!!

hongsii commented on src/main/java/domain/lottonumber/LottoGame.java:
```diff
@@ -0,0 +1,25 @@
+ * LottoNumbers를 추가하는 메서드와 결과값 생성 메서드를 가진다.
+ */
+public class LottoGame {
+    List<LottoTicket> lottoNumbers = new ArrayList<>();
```
전체적으로 접근제어자 설정이 default로 되어 있으니 바꿔보면 좋을 것 같습니다.

include42 replied:
지적 감사드립니다!! 리팩토링 진행하면서 자꾸 놓쳐서 주의해야겠습니다.

hongsii commented on src/main/java/domain/lottonumber/LottoGame.java:
```diff
@@ -0,0 +1,25 @@
+        lottoNumbers.add(lotto);
+    }
+
+    public void makeResult(LottoResult lottoResult, LottoWinner lottoWinner) {
```
이 메서드에서 `LottoResult` 를 생성하면 어떨까요? 이 클래스를 호출하는 곳에서 매번 인스턴스를 생성할 필요가 없을 듯 합니다.
반환형이 없는 메서드는 테스트하기 힘들기 때문에 가능하면 입출력이 있는 메서드로 작성하려고 노력해보시면 좋을 것 같아요~

include42 replied:
말씀주신 대로 하니 훨씬 깔끔하네요!! 감사합니다ㅎㅎ

hongsii commented on src/main/java/domain/lottonumber/LottoNumber.java:
```diff
@@ -0,0 +1,29 @@
+public class LottoNumber implements Comparable<LottoNumber> {
+    private static final String ERROR_BOUND_MESSAGE = "1부터 45 사이의 숫자만 입력 가능합니다.";
+    private static final int LOTTO_UNDER_BOUND = 1;
+    private static final int LOTTO_UPPER_BOUND = 45;
```
여기도 띄어쓰기를 넣어주면 좋을 듯 해요~ 전체적으로 일정한 컨벤션을 맞추려고 노력해보시면 좋을 듯 합니다.

include42 replied:
리팩토링을 진행하면서 컨벤션이나 접근제어자를 많이 놓쳤네요. 이런 부분을 꼼꼼히 보는 습관 가져야겠습니다!!

hongsii reviewed (changes requested):
피드백 반영을 잘 해주셨네요 👍 
사소한 피드백 몇 가지 드렸으니, 반영해주세요!

include42 committed:
[refactor] 개선사항 지적에 따른 코드 리팩토링

LottoController makeWinnerNumber와 makeResult 로직을 하나로 합침
UserLottoGenerator 그때그때 객체를 생성하는 식으로 불변성 보장
LottoGame makeResult가 LottoResult 인스턴스를 반환하도록 재설계
LottoNumber 컨벤션에 맞춘 코드 수정

include42 committed:
[refactor] LottoRank의 calculateRank 로직 단순화

LottoRank 기존 calculateRank의 복잡한 로직을 단순화함.

include42 committed:
[refactor] 컨벤션 준수 및 가독성을 위한 코드 수정

LottoController 가독성을 위해 메소드 위치 변경
LottoNumberFactory 컨벤션 준수를 위한 코드 리팩토링
LottoRank 컨벤션 준수를 위한 코드 리팩토링

hongsii commented on src/main/java/domain/lottoresult/LottoRank.java:
```diff
@@ -0,0 +1,54 @@
+     */
+    public static LottoRank calculateRank(int hitCount, boolean bonus) {
+        return Arrays.stream(LottoRank.values())
+                .filter(rank -> rank.hitCount == hitCount && (!rank.bonus || bonus))
```
주석이 직관적일 수 있지만, 아래와 같이 코드로 의미를 드러내보아도 좋을 것 같습니다.

``` java
private boolean checkBonus(boolean bonus) {
    boolean skipToCheck = !this.bonus
    return skipToCheck || bonus
}
```

include42 replied:
감사합니다!! 조건부분을 분리하는게 더 직관적일수 있겠군요

hongsii reviewed (approved):
피드백 반영을 잘해주셨네요 👍 
머지하겠습니다. 다음 단계 진행해주세요!
